### ISSUE TITLE
False positive on single_char_pattern with serde_json::Value.find(_)

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#830]
```
fn foo(v: &serde_json::Value) {
    let _ = v.find("x")...
}
```
will trigger the lint, however `Value` cannot `find` single characters for some reason. (Found in https://github.com/Thinkofname/steven-rust)

**Comment by oli-obk** [comment]
That's not really a false positive, but a hint that `Value` should be fixed. How about we lint on implementing a `PATTERN_METHODS` method on a type, but it has the wrong signature? `serde` already runs clippy, so we'd catch that there.

**Comment by llogiq** [comment]
Yes, I am aware that serde's Value implementation is to blame. However, the user likely isn't – and as long as we don't go around fixing incomplete `PATTERN_METHODS` implementations, to them this is a false positive, regardless if clippy is doing the right thing. So we should document that somehow.

Linting incomplete implementations would be a good way forward. I'm doing something similar with `len_zero` and `len_without_is_empty`.

**Comment by oli-obk** [comment]
can we also check if the method's generics are what we think they should be and not lint in that case?

Basically:

1. is a generic method?
2. has just one generic argument?
3. generic argument is also the type of the passed argument?
4. generic argument requires `Pattern` trait?
5. lint

**Comment by jugglerchris** [comment]
There's a similar false positive on `foo.contains("X")` where `foo` is a `HashSet<String>`.  `HashSet<T>::contains` takes a `T`, not a `Pattern`.

**Comment by mcarton** [comment]
Yes, the current lint ignores the type of `foo`. I’ve looked at that, but finding if the method generics at that point is tricky.

**Comment by shssoichiro** [comment]
When this was originally implemented, [I had also suggested](https://github.com/Manishearth/rust-clippy/issues/650#issuecomment-183820315) searching for `Pattern` trait as an argument. At that time it was deemed too difficult, so this lint was implemented with a list of method names to check on. Hence this issue.

I think improving this lint to search for `Pattern` arguments rather than a list of method names would be the safest/most complete fix, but the consensus seems to be that implementing that would be difficult. If I have free time maybe I can go back in and give this a second shot.
